solutionbox/inception/datalab_solutions/inception/_predictor.py

```diff
@@ -82,57 +59,181 @@ def predict(model_dir, image_files):
+    new_saver = tf.train.import_meta_graph(os.path.join(self._model_dir, 'export.meta'))
+    new_saver.restore(self._session, os.path.join(self._model_dir, 'export'))
+    init_op = tf.get_collection(tf.contrib.session_bundle.constants.INIT_OP_KEY)[0]
+    self._session.run(init_op)
+    self._tf_inputs = json.loads(tf.get_collection('inputs')[0])
+    self._tf_outputs = json.loads(tf.get_collection('outputs')[0])
+
+  def finish_bundle(self, context):
+    if self._session is not None:
+      self._session.close()
+
+  def process(self, context):
+    import collections
+    
+    element = context.element
+    image_urls = [x['image_url'] for x in element]
+    targets = None
+    if 'label' in element[0] and element[0]['label'] is not None:
+      targets = [x['label'] for x in element]
+
+    feed_dict = collections.defaultdict(list)
+    feed_dict[self._tf_inputs['image_bytes']] = [x['image_bytes'] for x in element]
+    feed_dict[self._tf_inputs['key']] = image_urls
+    predictions, labels, scores = self._session.run(
+        [self._tf_outputs['prediction'], self._tf_outputs['labels'], self._tf_outputs['scores']],
+        feed_dict=feed_dict)
+    if targets is not None:
+      yield zip(image_urls, targets, predictions, labels, scores)
+    else:
+      yield zip(image_urls, predictions, labels, scores)
+
+
+class ProcessResultsDoFn(beam.DoFn):
+  """A DoFn that process prediction results."""
+
+  def process(self, context):
+    element = context.element
+    target = None
+    if len(element) == 5:
+      image_url, target, prediction, labels, scores = element
+    else:
+      image_url, prediction, labels, scores = element
+    labels = list(labels)
+    predicted_prob = scores[labels.index(prediction)]
+    out_element = {
+      'image_url': image_url,
+      'predicted': prediction,
+      # Convert to float from np.float32 because BigQuery Sink can only handle intrinsic types.
+      'predicted_prob': float(predicted_prob)  
+    }
+    if target is not None:
+      target_prob = scores[labels.index(target)] if target in labels else 0.0
+      out_element['target_prob'] = float(target_prob)
+      out_element['target'] = target
+    yield out_element
+
+
+class MakeCsvLineDoFn(beam.DoFn):
+  """A DoFn that makes CSV lines out of prediction results."""
+
+  def process(self, context):
+    import csv
+    import StringIO
+
+    element = context.element
+    line = StringIO.StringIO()
+    if len(element) == 5:    
+      csv.DictWriter(line, 
+          ['image_url', 'target', 'predicted', 'target_prob', 'predicted_prob']).writerow(element)
+    else:
+      csv.DictWriter(line, ['image_url', 'predicted', 'predicted_prob']).writerow(element)   
+    yield line.getvalue()
+
+
+def configure_pipeline(p, dataset, model_dir, output_csv, output_bq_table):
+  """Configures a dataflow pipeline for batch prediction."""
+
+  data = _util.get_sources_from_dataset(p, dataset, 'predict')
+  schema = _util.get_schema_from_dataset(dataset)
```

Comment:
Why not dataset.schema? Oh, you do some error checking. Still looks odd that you have to wrap "get schema" and our dataset classes cannot do this.